pkg/volumes/alicloud/volumes.go

```diff
@@ -0,0 +1,318 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package alicloud
+
+import (
+	"fmt"
+	"os"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	alicloud "github.com/aliyun/alibaba-cloud-sdk-go/sdk"
+	"github.com/aliyun/alibaba-cloud-sdk-go/sdk/auth/credentials"
+	"github.com/aliyun/alibaba-cloud-sdk-go/sdk/requests"
+	"github.com/aliyun/alibaba-cloud-sdk-go/services/ecs"
+	"github.com/golang/glog"
+	"kope.io/etcd-manager/pkg/volumes"
+)
+
+const (
+	// We could query at most 50 disks at a time on Alicloud
+	PageSizeLarge = 50
+)
+
+type AlicloudVolumes struct {
+	mutex sync.Mutex
+
+	matchTagKeys []string
+	matchTags    map[string]string
+	nameTag      string
+	clusterName  string
+
+	deviceMap  map[string]string
+	ecs        *ecs.Client
+	instanceID string
+	localIP    string
+	metadata   *ECSMetadata
+	az         string
+	region     string
+}
+
+var _ volumes.Volumes = &AlicloudVolumes{}
+
+// NewAlicloudVolumes returns a new aws volume provider
+func NewAlicloudVolumes(clusterName string, volumeTags []string, nameTag string) (*AlicloudVolumes, error) {
+	a := &AlicloudVolumes{
+		clusterName: clusterName,
+		deviceMap:   make(map[string]string),
+		matchTags:   make(map[string]string),
+		nameTag:     nameTag,
+	}
+
+	for _, volumeTag := range volumeTags {
+		tokens := strings.SplitN(volumeTag, "=", 2)
+		if len(tokens) == 1 {
+			a.matchTagKeys = append(a.matchTagKeys, tokens[0])
+		} else {
+			a.matchTags[tokens[0]] = tokens[1]
+		}
+	}
+
+	a.metadata = NewECSMetadata()
+
+	var err error
+
+	a.region, err = a.metadata.GetMetadata("region-id")
+	if err != nil {
+		return nil, fmt.Errorf("error querying ecs metadata (for region): %v", err)
+	}
+
+	a.az, err = a.metadata.GetMetadata("zone-id")
+	if err != nil {
+		return nil, fmt.Errorf("error querying ecs metadata (for az): %v", err)
+	}
+
+	a.instanceID, err = a.metadata.GetMetadata("instance-id")
+	if err != nil {
+		return nil, fmt.Errorf("error querying ecs metadata (for instance-id): %v", err)
+	}
+
+	a.localIP, err = a.metadata.GetMetadata("private-ipv4")
+	if err != nil {
+		return nil, fmt.Errorf("error querying ecs metadata (for private-ipv4): %v", err)
+	}
+
+	ramRole, err := a.metadata.GetMetadata("ram/security-credentials/")
+	if err != nil {
+		return nil, fmt.Errorf("error querying ecs metadata (for RAM role): %v", err)
+	}
+
+	config := alicloud.NewConfig()
+
+	credential := credentials.NewEcsRamRoleCredential(ramRole)
+	a.ecs, err = ecs.NewClientWithOptions(a.region, config, credential)
```

Comment:
Looks like we aren't checking `err` here.

Reply:
Fixed and rebased. Thanks.